engine/common/follower/engine.go

```diff
@@ -117,122 +141,149 @@ func New(
 		}
-	})
+	}
 }
 
-// ProcessLocal processes an event originating on the local node.
-func (e *Engine) ProcessLocal(event interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(e.me.NodeID(), event)
-	})
-}
+// processQueuedBlocks processes any available messages until the message queue is empty.
+// Only returns when all inbound queues are empty (or the engine is terminated).
+// No errors are expected during normal operation. All returned exceptions are potential
+// symptoms of internal state corruption and should be fatal.
+func (e *Engine) processQueuedBlocks() error {
+	for {
+		msg, ok := e.pendingBlocks.Pop()
+		if ok {
+			in := msg.(inboundBlock)
+			err := e.processBlockProposal(in.originID, in.block)
+			if err != nil {
+				return fmt.Errorf("could not handle block proposal: %w", err)
+			}
+			e.engMetrics.MessageHandled(metrics.EngineFollower, metrics.MessageBlockProposal)
+			continue
+		}
 
-// Process processes the given event from the node with the given origin ID in
-// a blocking manner. It returns the potential processing error when done.
-func (e *Engine) Process(_ channels.Channel, originID flow.Identifier, event interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(originID, event)
-	})
+		// when there are no more messages in the queue, back to the processBlocksLoop to wait
+		// for the next incoming message to arrive.
+		return nil
+	}
 }
 
-func (e *Engine) process(originID flow.Identifier, input interface{}) error {
-	switch v := input.(type) {
-	case *messages.BlockResponse:
-		e.engMetrics.MessageReceived(metrics.EngineFollower, metrics.MessageBlockResponse)
-		defer e.engMetrics.MessageHandled(metrics.EngineFollower, metrics.MessageBlockResponse)
-		e.unit.Lock()
-		defer e.unit.Unlock()
-		return e.onBlockResponse(originID, v)
-	case *events.SyncedBlock:
-		e.engMetrics.MessageReceived(metrics.EngineFollower, metrics.MessageSyncedBlock)
-		defer e.engMetrics.MessageHandled(metrics.EngineFollower, metrics.MessageSyncedBlock)
-		e.unit.Lock()
-		defer e.unit.Unlock()
-		return e.onSyncedBlock(originID, v)
-	case *messages.BlockProposal:
-		e.engMetrics.MessageReceived(metrics.EngineFollower, metrics.MessageBlockProposal)
-		defer e.engMetrics.MessageHandled(metrics.EngineFollower, metrics.MessageBlockProposal)
-		e.unit.Lock()
-		defer e.unit.Unlock()
-		return e.onBlockProposal(originID, v)
-	default:
-		return fmt.Errorf("invalid event type (%T)", input)
+// onBlockProposal performs processing of incoming block by pushing into queue and notifying worker.
+func (e *Engine) onBlockProposal(originID flow.Identifier, proposal *messages.BlockProposal) {
+	e.engMetrics.MessageReceived(metrics.EngineFollower, metrics.MessageBlockProposal)
+	// queue as proposal
+	in := inboundBlock{originID, proposal}
+	if e.pendingBlocks.Push(in) {
+		e.pendingBlocksNotifier.Notify()
 	}
 }
 
+// onSyncedBlock performs processing of incoming block by pushing into queue and notifying worker.
```

Comment:
Missing error docs, but I imagine we'll be cleaning this up as part of https://github.com/dapperlabs/flow-go/issues/6424